fastscanner/services/indicators/service.py

```diff
@@ -136,18 +138,90 @@ async def handle(self, channel_id: str, data: dict[Any, Any]) -> None:
             if "timestamp" not in data:
                 logger.warning(f"Missing timestamp in message from {channel_id}")
                 return
+
             ts = pd.to_datetime(int(data["timestamp"]), unit="ms", utc=True).tz_convert(
                 LOCAL_TIMEZONE_STR
             )
             new_row = pd.Series(data, name=ts)
 
-            for indicator in self._indicators:
-                new_row = await indicator.extend_realtime(self._symbol, new_row)
-
-            self._handler.handle(self._symbol, new_row)
+            await self._buffer.add(new_row)
 
         except Exception as e:
             logger.error(
                 f"[Handler Error] Failed processing message from {channel_id}: {e}"
             )
+
+
+class CandleBuffer:
+    def __init__(
+        self,
+        symbol: str,
+        freq: str,
+        handler: SubscriptionHandler,
+        indicators: list[Indicator],
+    ):
+        self.symbol = symbol
+        self.freq = freq
+        self.handler = handler
+        self.indicators = indicators
+        self.buffer: dict[pd.Timestamp, pd.Series] = {}
+        self.start_time: pd.Timestamp | None = None
+        self.expected_count = self._get_expected_count()
+        self.timeout_task: asyncio.Task | None = None
+
+    def _get_expected_count(self) -> int:
+        n, unit = split_freq(self.freq)
+        return n if unit == "min" else n * 60
+
+    def _get_interval_start(self, ts: pd.Timestamp) -> pd.Timestamp:
+        freq_minutes = self._get_expected_count()
+        minute_group = (ts.minute // freq_minutes) * freq_minutes
+        return ts.replace(minute=minute_group, second=0, microsecond=0)
```

Comment:
```suggestion
    def _get_interval_start(self, ts: pd.Timestamp) -> pd.Timestamp:
        return ts.floor(self._freq)
```
As the function is so simple, we don't need to have a _get_interval_start.